### PR TITLE
Monitoring

### PR DESCRIPTION
**Pull request by byucesoy**
**Move exception_to_hash to Util**
Previously, only dispatcher was using this helper function, but it is useful in
other contexes such as in monitoring. Thus, we are moving it to Util.

**Add monitor process**
We are adding a new executable for health monitor. This process would fetch the
list of resources that needs to be monitored and checks their health in a loop.

Here are the things we specifically cared about while implementing this;
- We are decoupling health checks from the resource's nexus progs. This allows
us to perform very frequent health checks independent from the nexus' schedule.
It also does not cause unnecessary lock holding while doing the health checks.
Finally this allows us to independently scale monitoring process.
- We use persistent SSH sessions for monitoring. This ensures very fast health
check cycles. Moreover, we even use SSH tunnels to communicate with the target
software directly, thus avoiding many expensive system calls and forks, though
example of such communication is not available in this commit. See the next one
for an example.
- Each resource decides how the persistent session must be created and how the
check itself must be performed. It does this by defining two functions called;
init_health_monitor_session and check_pulse. Monitoring process creates the
session using init_health_monitor_session and passes it to check_pulse. Primary
contract between the monitor and the resource is that the session needs to be
based on SSH. In a similar fashion, resource can decide how frequently it needs
to be monitored by defining monitoring_interval function.

**Add helper to sshable to start fresh SSH sessions**
In the context of monitoring, we want a fresh SSH session that is not taken
from the thread's cache, because we are calling session.process ourselves and
doing that interferes with regular command execution. To prevent that, we are
creating a fresh SSH session for the resources we want to monitor.

**Add monitoring for PosgresServer**
For running SELECT 1 queries without forking postmaster and without running
expensive system calls, we are using persistent sessions over SSH tunnel. This
is done by forwarding local UNIX socket to the Postgres socket in the target
server.

This is faster than running `psql -c 'SELECT 1'`, because it doesn't spawn new
process thus doesn't make lots of system calls. In my rough experiments using
sockets is 3-4x faster than running `psql -c 'SELECT 1'`.

This approach is also significantly reduces the memory allocations, which could
cause OOM problems when there is a high load on the system and make us declare
unavailability unnecessarily.

**Try to recover Postgres from unavailability**
Currently we just try to restart PostgreSQL process. In the future, this can be
extended to triggering failover for HA enabled databases or creating a standby
for non-HA databases.

One usual false positive is database not responding to queries during the crash
recovery. In such cases, triggering restart is counter-productive as it would
undo some part of crash recovery. To prevent that, we check the logs and not
trigger restart if the database is in crash recovery.

**Allow using sockets in MinIO client**
This allows us to use sockets for faster and more efficient communication with
the web server.

**Move connection_string to individual MinioServers**
MinioServer itself also needs to know its own connection string. Instead of
duplicating the code, we move connection_string to MinioServer model and the
MinioCluster refers to the connection_string functin in the MinioServer.

**Add monitoring for MinioServer**
We are using admin_info endpoint provided by MinIO, which returns status of
each server and their drives. Similar to PostgresServer, we are using UNIX
sockets for the communication, which avoids doing unnecessary system calls and
memory allocations.

**Try to recover MinIO from unavailability**
Currently only attempt for recovery is restarting the MinIO service, which
seemed to help in few cases. In the future, we might automate VM replacement.

### CODE REVIEW

File: bin/monitor [hunk 1]
```diff
@@ -3,7 +3,7 @@
```
Comment (fdr) [inline bin/monitor:3]: re, commit message:

```

This approach is also more memory efficient. Memory efficiency is especially
important during high loads. Process forking at those scenarios can cause OOM
errors, thus causing us to unnecessarily declare unavailability.
```

I'd reword "efficiency" (which tends to be read mostly about size, where it ends up being somewhat better for lack of an extra shell/bash and `psql`) and perhaps itemize (much) reduced frequency of memory mapping in the operating system.  It is this memory mapping that can fail with ENOMEM. The difference in magnitude in amount of new memory maps is likely thousands to one (as simple queries in a monitoring loop will do a few small pallocs in the same virtual memory region over and over, not running `mmap` nor `brk`/`sbrk`), the largest magnitude effect by far.

I am not sure about the optimal wording here, you could say "memory allocations," but don't care about most (small) memory allocations e.g. palloc/malloc as such, since they tend to grab a bunch of virtual memory from the OS (which is the part that fails) and then do bookkeeping inside there. Perhaps while a bit ambiguous, it more quickly gets the reader's head in the right place.

File: model/minio/minio_server.rb [hunk 2]
```diff
@@ -36,4 +38,42 @@ def minio_volumes
       pool.volumes_url
     end.join(" ")
   end
+
+  def connection_string
+    "http://#{vm.ephemeral_net4}:9000"
+  end
+
+  def init_health_monitor_session
+    socket_path = File.join(`pwd`.chomp, "health_monitor_sockets", "ms_#{vm.ephemeral_net6.nth(2)}")
```
Comment (fdr) [inline model/minio/minio_server.rb:47]:
```suggestion
    socket_path = File.join(Dir.pwd, "health_monitor_sockets", "ms_#{vm.ephemeral_net6.nth(2)}")
```

File: model/postgres/postgres_server.rb [hunk 3]
```diff
@@ -73,4 +75,37 @@ def configure_hash
...
+
+    if pulse[:reading] == "down" && pulse[:reading_rpt] > 5 && Time.now - pulse[:reading_chg] > 30
+      incr_checkup
+    end
+
+    pulse
+  end
+
+  def health_monitor_socket_path
+    @health_monitor_socket_path ||= File.join(`pwd`.chomp, "health_monitor_sockets", "pg_#{vm.ephemeral_net6.nth(2)}")
```
Comment (fdr) [inline model/postgres/postgres_server.rb:109]:
```suggestion
    @health_monitor_socket_path ||= File.join(Dir.pwd, "health_monitor_sockets", "pg_#{vm.ephemeral_net6.nth(2)}")
```

**Review by fdr — APPROVED** [review]
Very nice derivation, patch by patch. It made review far, far easier.

File: model.rb [hunk 4]
```diff
@@ -118,6 +118,20 @@ def redacted_columns
   end
 end
 
+module HealthMonitorMethods
```
Comment (enescakir) [inline model.rb:121]: I'm not sure if this aligns with the Ruby programming philosophy, but we could add stub methods to prompt developers to implement them in the resources.

```ruby
    def init_health_monitor_session
      raise NoMethodError
    end

    def check_pulse(session:, previous_pulse:)
      raise NoMethodError
    end
```

File: prog/postgres/postgres_server_nexus.rb [hunk 5]
```diff
@@ -179,18 +179,16 @@ def before_run
...
     when_initial_provisioning_set? do
-      hop_restart
+      decr_initial_provisioning
+      bud self.class, frame, :restart
     end
+
     hop_wait
   end
 
   label def restart
```
[the remark this answers is not in the document]
Comment (byucesoy) [reply prog/postgres/postgres_server_nexus.rb:189]: Moved them to after the destroy label. Another identifier might make sense for the future. Maybe "sink" as a borrowed term from automata theory.

File: model.rb (hunk 4 above)
Comment (byucesoy) [reply model.rb:121]: I don't know if Ruby philosophy says anything about this, but currently if we call `init_health_monitor_session` on a resource that doesn't implement it, we already get NoMethodError as this is what Ruby does by default.

File: prog/minio/minio_server_nexus.rb [hunk 6]
```diff
@@ -104,4 +122,22 @@ def cluster
...
+    client = Minio::Client.new(
+      endpoint: minio_server.connection_string,
+      access_key: minio_server.cluster.admin_user,
+      secret_key: minio_server.cluster.admin_password
+    )
+
+    server_data = JSON.parse(client.admin_info.body)["servers"].find { _1["endpoint"] == "#{minio_server.vm.ephemeral_net4}:9000" }
+    server_data["state"] == "online"
+  rescue
+    false
```
[the remark this answers is not in the document]
Comment (byucesoy) [reply prog/minio/minio_server_nexus.rb:141]: Yes, VM, Host, DNS monitoring will come later on. I didn't want to add them to this PR as its scope gets bigger and bigger.

File: model/minio/minio_server.rb [hunk 7]
```diff
@@ -40,4 +42,38 @@ def minio_volumes
...
+    FileUtils.rm_rf(socket_path)
+    FileUtils.mkdir_p(socket_path)
+
+    ssh_session = vm.sshable.start_fresh_session
+    ssh_session.forward.local(UNIXServer.new(File.join(socket_path, "health_monitor_socket")), private_ipv4_address, 9000)
+    {
+      ssh_session: ssh_session,
+      minio_client: Minio::Client.new(
+        endpoint: connection_string,
+        access_key: cluster.admin_user,
```
[the remark this answers is not in the document]
Comment (byucesoy) [reply model/minio/minio_server.rb:55]: after offline discussion, we decided to not do this as in this scope admin_user is already reachable, thus adding new user doesn't add any protection.

File: bin/monitor (hunk 1 above)
Comment (byucesoy) [reply bin/monitor:3]: Updated the commit message.
